File: server/src/main/java/org/queue/consumer/TopicCount.scala
/*
 * Copyright 2010 LinkedIn
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.queue.consumer

import com.google.gson.Gson
import org.slf4j.LoggerFactory

import scala.collection.{Map, mutable}
import scala.jdk.CollectionConverters.MapHasAsScala

private[consumer] object TopicCount {
  private val logger = LoggerFactory.getLogger(getClass())
  val gson = new Gson()
  def constructTopicCount(consumerIdSting: String, jsonString : String) : TopicCount = {
    if("{  }".equals(jsonString)){
      return new TopicCount(consumerIdSting, gson.fromJson("", classOf[java.util.Map[String, Int]]))
    }

    var value :java.util.Map[String, Int] = null
    try {
      value = gson.fromJson(jsonString, classOf[java.util.Map[String, Int]])
      if(value == null){
        throw new RuntimeException("error constructing TopicCount : " + jsonString)
      }
    }catch {
      case e : Throwable =>
        logger.error("error parsing consumer json string " + jsonString, e)
        throw e
    }
    new TopicCount(consumerIdSting, value.asScala)
  }

}

private[consumer] class TopicCount(val consumerIdString: String, val topicCountMap: Map[String, Int]) {

  def getConsumerThreadIdsPerTopic(): Map[String, Set[String]] = {
    val consumerThreadIdsPerTopicMap = new mutable.HashMap[String, Set[String]]()
    for ((topic, nConsumers) <- topicCountMap) {
      val consumerSet = new mutable.HashSet[String]
      assert(nConsumers >= 1)
      for (i <- 0 until nConsumers)
        consumerSet += consumerIdString + "-" + i
      consumerThreadIdsPerTopicMap.put(topic, consumerSet.toSet)
    }
    consumerThreadIdsPerTopicMap
  }

  override def equals(obj: Any): Boolean = {
    obj match {
      case null => false
      case n: TopicCount => consumerIdString == n.consumerIdString && topicCountMap == n.topicCountMap
      case _ => false
    }
  }

  /**
   *  return json of
   *  { "topic1" : 4,
   *    "topic2" : 4
   *  }
   */
  def toJsonString() : String = {
    val builder = new StringBuilder
    builder.append("{ ")
    var i = 0
    for ( (topic, nConsumers) <- topicCountMap) {
      if (i > 0)
        builder.append(",")
      builder.append("\"" + topic + "\": " + nConsumers)
      i += 1
    }
    builder.append(" }")
    builder.toString
  }
}

File: server/src/main/java/org/queue/consumer/TopicCount.java
package org.queue.consumer;

/**
 * @author gaoyvfeng
 * @ClassName TopicCount
 * @description:
 * @datetime 2024年 05月 24日 10:01
 * @version: 1.0
 */
import com.google.gson.Gson;
import java.util.Map;
import java.util.Set;
import java.util.HashMap;
import java.util.HashSet;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class TopicCount {
    private static final Logger logger = LoggerFactory.getLogger(TopicCount.class);
    private final String consumerIdString;
    private final Map<String, Integer> topicCountMap;
    private static Gson gson = new Gson();

    public static TopicCount constructTopicCount(String consumerIdString, String jsonString) {
        if ("{  }".equals(jsonString)) {
            return new TopicCount(consumerIdString, gson.fromJson("", Map.class));
        }

        Map<String, Integer> value;
        try {
            value = gson.fromJson(jsonString, Map.class);
            if (value == null) {
                throw new RuntimeException("error constructing TopicCount: " + jsonString);
            }
        } catch (Throwable e) {
            logger.error("error parsing consumer json string " + jsonString, e);
            throw e;
        }
        // 待办：处理JSON转换问题
        return new TopicCount(consumerIdString, new HashMap<String, Integer>() {{ put("topic1", 1); }});
    }


    public TopicCount(String consumerIdString, Map<String, Integer> topicCountMap) {
        this.consumerIdString = consumerIdString;
        this.topicCountMap = topicCountMap;
    }

    public Map<String, Set<String>> getConsumerThreadIdsPerTopic() {
        Map<String, Set<String>> consumerThreadIdsPerTopicMap = new HashMap<>();
        for (Map.Entry<String, Integer> entry : topicCountMap.entrySet()) {
            Set<String> consumerSet = new HashSet<>();
            assert (entry.getValue() >= 1);
            for (int i = 0; i < entry.getValue(); i++) {
                consumerSet.add(consumerIdString + "-" + i);
            }
            consumerThreadIdsPerTopicMap.put(entry.getKey(), consumerSet);
        }
        return consumerThreadIdsPerTopicMap;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        TopicCount topicCount = (TopicCount) obj;
        return consumerIdString.equals(topicCount.consumerIdString) &&
                topicCountMap.equals(topicCount.topicCountMap);
    }

    /**
     * 返回形如
     * { "topic1" : 4,
     *   "topic2" : 4
     * }
     * 的JSON字符串。
     */
    public String toJsonString() {
        StringBuilder builder = new StringBuilder();
        builder.append("{ ");
        int i = 0;
        for (Map.Entry<String, Integer> entry : topicCountMap.entrySet()) {
            if (i++ > 0) builder.append(", ");
            builder.append("\"").append(entry.getKey()).append("\": ").append(entry.getValue());
        }
        builder.append(" }");
        return builder.toString();
    }
}
